src/runtime/cpu_device_api.cc

```diff
@@ -32,8 +35,20 @@ class CPUDeviceAPI final : public DeviceAPI {
     ptr = memalign(alignment, nbytes);
     if (ptr == nullptr) throw std::bad_alloc();
 #else
-    int ret = posix_memalign(&ptr, alignment, nbytes);
-    if (ret != 0) throw std::bad_alloc();
+#ifndef _WIN32
+    constexpr size_t _HugePage2MB_ = 1<<21;
+    if (nbytes >= _HugePage2MB_) {
+       int ret = posix_memalign(&ptr, _HugePage2MB_, nbytes);
+       if (ret != 0) throw std::bad_alloc();
+       if ((ret= madvise(ptr, nbytes, MADV_HUGEPAGE )) != 0 )
```

Comment:
if `nbytes` isn't divisable by 2M, will a portion of the memory still be allocated with huge page?
I remember you said this kind of huge allocation is only available in the latest linux kernel. do you know which version of linux starts to support something like this?

Reply:
Since 2.6.38.  I would say this is pretty safe to call for Linux.

Reply:
For the first question the answer is [yes](https://man7.org/linux/man-pages/man2/madvise.2.html), hence the check on L40.